src/complex_fluids/ComplexFluidForcing.cpp

```diff
@@ -0,0 +1,957 @@
+        {
+            hier_cc_data_ops->copyData(d_W_cc_idx, W_current_idx);
+        }
+        else
+        {
+            hier_cc_data_ops->linearSum(d_W_cc_idx, 0.5, W_current_idx, 0.5, W_new_idx);
+        }
+    }
+
+    HierarchyDataOpsManager<NDIM>* hier_data_ops_manager = HierarchyDataOpsManager<NDIM>::getManager();
+    Pointer<HierarchyDataOpsReal<NDIM, double>> hier_cc_data_ops =
+        hier_data_ops_manager->getOperationsDouble(d_W_cc_var, hierarchy);
+    if (d_sqr_root)
+    {
+        squareMatrix(d_W_scratch_idx, d_W_cc_var, hierarchy, data_time, initial_time, coarsest_ln, finest_ln);
+    }
+    else if (d_log_conform)
+    {
+        exponentiateMatrix(d_W_scratch_idx, d_W_cc_var, hierarchy, data_time, initial_time, coarsest_ln, finest_ln);
+    }
+    else
+    {
+        if (d_project_conform)
+        {
+            // TODO: This only projects the conformation tensor data owned by this class. Really, we need to project
+            // the conformation tensor owned by the advection diffusion integrator.
+            projectMatrix(d_W_cc_idx, d_W_cc_var, hierarchy, data_time, initial_time, coarsest_ln, finest_ln);
+        }
+        hier_cc_data_ops->copyData(d_W_scratch_idx, d_W_cc_idx);
+    }
+
+    typedef HierarchyGhostCellInterpolation::InterpolationTransactionComponent InterpolationTransactionComponent;
+    std::vector<InterpolationTransactionComponent> ghost_cell_components(1);
+    ghost_cell_components[0] = InterpolationTransactionComponent(d_W_scratch_idx,
+                                                                 "CONSERVATIVE_LINEAR_REFINE",
+                                                                 true,
+                                                                 "CONSERVATIVE_COARSEN",
+                                                                 "LINEAR",
+                                                                 false,
+                                                                 d_conc_bc_coefs,
+                                                                 NULL,
+                                                                 d_interp_type);
+    HierarchyGhostCellInterpolation ghost_fill_op;
+    ghost_fill_op.initializeOperatorState(ghost_cell_components, hierarchy);
+    ghost_fill_op.fillData(data_time);
+
+    // Check to ensure conformation tensor is positive definite
+    d_positive_def = true;
+    checkPositiveDefinite(d_W_scratch_idx, d_W_cc_var, data_time, initial_time);
+    int temp = d_positive_def ? 1 : 0;
+    d_positive_def = SAMRAI_MPI::maxReduction(temp) == 1 ? true : false;
+    plog << "Conformation tensor is " << (d_positive_def ? "SPD" : "NOT SPD") << "\n";
+    if (d_error_on_spd && !d_positive_def)
+    {
+        TBOX_ERROR("ERROR: \n "
+                   << "    CONFORMATION TENSOR IS NOT SPD! \n\n");
+    }
+
+    // Check max and min determinant
+    if (d_log_det)
+    {
+        d_max_det = 0.0;
+        d_min_det = 1.0e10;
```

Comment:
`std::numeric_limits<double>::max()` ?

Reply:
This is something I'm not sure about. Will this affect `max` and `min` MPI reductions?

Reply:
MPI reductions will be fine.

Practically speaking, 1e10 is also probably fine, but I think it is clearer to use `numeric_limits` for large values.